app/src/main/kotlin/pl/gov/mc/protego/bluetooth/BeaconIdManager.kt

```diff
@@ -0,0 +1,67 @@
+package pl.gov.mc.protego.bluetooth
+
+import android.os.Handler
+import pl.gov.mc.protego.bluetooth.beacon.BeaconId
+import pl.gov.mc.protego.bluetooth.beacon.BeaconIdAgent
+import pl.gov.mc.protego.bluetooth.beacon.BeaconIdLocal
+import pl.gov.mc.protego.bluetooth.beacon.BeaconIdRemote
+import timber.log.Timber
+import java.util.*
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+
+
+class BeaconIdManager : BeaconIdAgent {
+
+    private val beaconIdPart = AtomicInteger(0x0f)
+    private val handler = Handler()
+    private fun createNewBeaconId(): BeaconIdLocal {
+        val newBeaconId = BeaconIdLocal(
+            BeaconId(
+                byteArrayOf(
+                    0x00,
+                    0x01,
+                    0x02,
+                    0x03,
+                    0x04,
+                    0x05,
+                    0x06,
+                    0x07,
+                    0x08,
+                    0x09,
+                    0x0a,
+                    0x0b,
+                    0x0c,
+                    0x0d,
+                    0x0e,
+                    beaconIdPart.getAndIncrement().toByte()
+                ), ProteGoManufacturerDataVersion
+            ), Date(System.currentTimeMillis() + TimeUnit.MINUTES.toMillis(3))
+        )
+        handler.postDelayed({ currentBeaconId = createNewBeaconId() }, newBeaconId.expirationDate.time - System.currentTimeMillis())
+        return newBeaconId
+    }
+
+    private var listeners = setOf<BeaconIdAgent.Listener>()
+
+    private var currentBeaconId = createNewBeaconId()
+        set (value) {
+            listeners.forEach { it.useBeaconId(value) }
+            field = value
+        }
+
+    override fun getBeaconId(): BeaconIdLocal? = currentBeaconId
+
+    override fun registerListener(listener: BeaconIdAgent.Listener) {
+        listeners = listeners + listener
```

Comment:
Dodanie każdego elementu powoduje tworzenie nowego setu, kopiowanie do niego zawartości obecnego setu i dodanie nowej wartosci. Dlaczego nie jest to po prostu HashSet?

Reply:
Po zmianie na MutableSet:
```suggestion
        listeners.add(listener)
```

Reply:
Tak, dodawanie słuchaczy jest stosunkowo rzadką czynnością a pozwoli na uniknięcie potencjalnych `ConcurrentModificationException`

Reply:
Rozumiem, skoro wielowątkowość wchodzi w grę to ma sens. Ewentualnie można rozważyć `ConcurrentHashMap.newKeySet<BeaconIdAgent.Listener>()`

Reply:
inne potencjalne rozwiązanie to `CopyOnWriteArrayList` która jest thread safe

Reply:
Jeśli `BeaconIdManager` ma być thread-safe, to modyfikacja pola `listeners` powinna być synchronizowana. W innym razie, możliwa jest poniższa sytuacja:
1. BeaconIdManager jest w początkowym stanie.
2. Dwa wątki równolegle wywołują metodę `registerListener`
3. Każdy wątek niezależnie ustawia pole listeners
Efekt: nie mamy pewności, czy oba listenery zostaną finalnie dodane.

Rozwiązaniem tego byłoby dodanie dedykowanego lock'a dla modyfikacji pola `listeners` i wykorzystanie go w nowej metodzie jak `addListener(BeaconIdAgent.Listener)`. W ten sposób będzie pewność, że żaden listener nie zgubi się gdzieś wśród wielu równoległych wywołań rejestracji listenera.

(PS: Nie należę do projektu, po prostu znalazłem na jakiejś stronie link do repo i zainteresowałem się dyskusją :D )

Reply:
> inne potencjalne rozwiązanie to `CopyOnWriteArrayList` która jest thread safe

Tylko, że ArrayList nie zapewnia unikalności elementów tak jak set.

Reply:
`CopyOnWriteArraySet`?

Reply:
za https://stackoverflow.com/a/6720658 chyba `CopyOnWriteArraySet` jest dobrym wyborem